pkg/crc/machine/kubeconfig.go

```diff
@@ -0,0 +1,150 @@
+)
+
+const (
+	adminContext     = "crc-admin"
+	developerContext = "crc-developer"
+)
+
+func eventuallyWriteKubeconfig(ocConfig oc.Config, clusterConfig *ClusterConfig) error {
+	if err := errors.RetryAfter(60, func() error {
+		status, err := cluster.GetClusterOperatorStatus(ocConfig, "authentication")
+		if err != nil {
+			return &errors.RetriableError{Err: err}
+		}
+		if isReady(status) {
+			return nil
+		}
+		return &errors.RetriableError{Err: goerrors.New("cluster operator authentication not ready")}
+	}, 2*time.Second); err != nil {
+		logging.Warn("Skipping the kubeconfig update. Cluster operator authentication still not ready after 2min.")
+	} else if err := writeKubeconfig(clusterConfig); err != nil {
+		return err
+	}
+	return nil
+}
+
+func isReady(status *cluster.Status) bool {
+	return status.Available && !status.Progressing && !status.Degraded && !status.Disabled
+}
+
+func writeKubeconfig(clusterConfig *ClusterConfig) error {
+	kubeconfig := getGlobalKubeConfigPath()
+	dir := filepath.Dir(kubeconfig)
+	if err := os.MkdirAll(dir, 0755); err != nil {
+		return err
+	}
+	// Make sure .kube/config exist if not then this will create
+	_, _ = os.OpenFile(kubeconfig, os.O_RDONLY|os.O_CREATE, 0600)
+
+	ca, err := certificateAuthority(clusterConfig)
+	if err != nil {
+		return err
+	}
+	host, err := hostname(clusterConfig.ClusterAPI)
+	if err != nil {
+		return err
+	}
+
+	cfg, err := clientcmd.LoadFromFile(kubeconfig)
+	if err != nil {
+		return err
+	}
+	cfg.Clusters[host] = &api.Cluster{
+		Server:                   clusterConfig.ClusterAPI,
+		CertificateAuthorityData: ca,
+	}
+
+	if err := addContext(cfg, clusterConfig, ca, adminContext, "kubeadmin", clusterConfig.KubeAdminPass); err != nil {
+		return err
+	}
+	if err := addContext(cfg, clusterConfig, ca, developerContext, "developer", "developer"); err != nil {
+		return err
+	}
+
+	if cfg.CurrentContext == "" {
+		cfg.CurrentContext = adminContext
+	}
+
+	return clientcmd.WriteToFile(*cfg, kubeconfig)
+}
+
+func certificateAuthority(clusterConfig *ClusterConfig) ([]byte, error) {
+	bin, err := ioutil.ReadFile(clusterConfig.KubeConfig)
+	if err != nil {
+		return nil, err
+	}
+	builtin, err := clientcmd.Load(bin)
+	if err != nil {
+		return nil, err
+	}
+	cluster, ok := builtin.Clusters["crc"]
+	if !ok {
+		return nil, fmt.Errorf("crc cluster not found in kubeconfig %s", clusterConfig.KubeConfig)
+	}
+	return cluster.CertificateAuthorityData, nil
+}
+
+func hostname(clusterAPI string) (string, error) {
+	p, err := url.Parse(clusterAPI)
+	if err != nil {
+		return "", err
+	}
+	return p.Host, nil
+}
+
+func addContext(cfg *api.Config, clusterConfig *ClusterConfig, ca []byte, context, username, password string) error {
+	host, err := hostname(clusterConfig.ClusterAPI)
+	if err != nil {
+		return err
+	}
+	token, err := tokencmd.RequestToken(&restclient.Config{
+		Host: clusterConfig.ClusterAPI,
+		TLSClientConfig: restclient.TLSClientConfig{
+			CAData: ca,
```

Comment:
@guillaumerose This makes the call from the host to the VM, that means host dns should be configured and working to resolve the `clusterConfig.ClusterAPI` hostname, I was looking into a way to get token using curl which we can execute directly inside the VM and get token.

```
$ curl -u developer:developer -kv -H "X-CSRF-Token: 1" 'https://oauth-openshift.apps-crc.testing/oauth/authorize?client_id=openshift-challenging-client&response_type=token' 2>&1 | grep -oP "access_token=\K[^&]*"
```

Reply:
It's now using a custom dialer that always connect to the VM IP. It skips the DNS resolution.